block_priofees/src/stats_calculation.rs

```diff
@@ -0,0 +1,118 @@
+use crate::rpc_data::{FeePoint, PrioFeesStats};
+use itertools::Itertools;
+use std::collections::HashMap;
+
+pub fn calculate_supp_stats(
+    // Vec(prioritization_fees, cu_consumed)
+    prio_fees_in_block: &Vec<(u64, u64)>,
+) -> PrioFeesStats {
+    let mut prio_fees_in_block = if prio_fees_in_block.is_empty() {
+        vec![(0, 0)]
+    } else {
+        prio_fees_in_block.clone()
+    };
+    // sort by prioritization fees
+    prio_fees_in_block.sort_by(|a, b| a.0.cmp(&b.0));
+
+    // get stats by transaction
+    let dist_fee_by_index: Vec<FeePoint> = (0..=100)
+        .step_by(5)
+        .map(|p| {
+            let prio_fee = if p == 100 {
+                prio_fees_in_block.last().unwrap().0
+            } else {
+                let index = prio_fees_in_block.len() * p / 100;
+                prio_fees_in_block[index].0
+            };
+            FeePoint {
+                p: p as u32,
+                v: prio_fee,
+            }
+        })
+        .collect_vec();
+
+    // get stats by CU
+    // e.g. 95 -> 3000
+    let mut dist_fee_by_cu: HashMap<i32, u64> = HashMap::new();
+    let cu_sum: u64 = prio_fees_in_block.iter().map(|x| x.1).sum();
+    let mut agg: u64 = 0;
+    for (prio, cu) in prio_fees_in_block {
+        agg += cu;
+        for p in (0..=100).step_by(5) {
```

Comment:
This algorithm is suboptimal, it is O(M*N) complex which can be converted to O(M) + O(N) by creating an array of increasing CU after executing the transaction and then in another pass comparing with percentage.

Reply:
updated